File: Semestru 4/MPP/lab/.net/lab3/lab3/Persistence/repo/implementation/ChildDbRepository.cs
using Model.app.domain;
using Persistence.app.repo.@interface;
using Persistence.app.utils;
using log4net;
using System.Data.SQLite;

namespace Persistence.app.repo.implementation
{
	public class ChildDbRepository : IChildRepository
	{
		private CommonUtils Utils;
		private static readonly ILog _logger = LogManager.GetLogger(typeof(ChildDbRepository));

		public ChildDbRepository ()
		{
			_logger.Info("Initialized ChildDbRepository.");
			this.Utils = new CommonUtils();
		}

		public Child Create(Child e)
		{
			_logger.Info($"Saving the child: {e.Name}, {e.Cnp}");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				using (var statement = new SQLiteCommand("insert into children (name, cnp) values (@name, @cnp)", connection))
				{
					statement.Parameters.Clear();
					statement.Parameters.AddWithValue("@name", e.Name);
					statement.Parameters.AddWithValue("@cnp", e.Cnp);
					int rows = statement.ExecuteNonQuery();
					_logger.Info($"Affected rows: {rows}");
					return (Child) new Child(e.Name, e.Cnp).SetId((int)connection.LastInsertRowId);
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
                Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public Child? Delete(int id)
		{
			throw new NotImplementedException();
		}

		public IEnumerable<Child> GetAll()
		{
			_logger.Info($"Searching for all children ");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				List<Child> children = new List<Child>();
				using (var query = new SQLiteCommand("select * from children", connection))
				{
					using (var reader = query.ExecuteReader())
					{
						while (reader.Read())
							children.Add(
								(Child)new Child(
									reader.GetString(reader.GetOrdinal("name")),
									reader.GetString(reader.GetOrdinal("cnp"))
								).SetId(reader.GetInt32(reader.GetOrdinal("id")))
						   );
						return children;
					}
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null!;
		}

		public IEnumerable<Child> GetAllByAge(int age)
		{
			throw new NotImplementedException();
		}

		public Child? GetById(int id)
		{
			_logger.Info($"Searching for the child with id: {id}");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				using (var query = new SQLiteCommand("select * from children where id = @id", connection))
				{
					query.Parameters.Clear();
					query.Parameters.AddWithValue("@id", id);
					using (var reader = query.ExecuteReader())
					{
						if (reader.Read())
							return (Child)new Child(
									reader.GetString(reader.GetOrdinal("name")),
									reader.GetString(reader.GetOrdinal("cnp"))
								).SetId(reader.GetInt32(reader.GetOrdinal("id")));
						return null;
					}
				}
			} catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null;
		}

		public Child? Update(Child e)
		{
			_logger.Info($"Updating the child: {e.Name}, {e.Cnp}");
			var connection = this.Utils.GetConnection();
			try
			{
				_logger.Info("Succesfully connected to the db.");
				if (connection.State != System.Data.ConnectionState.Open)
					connection.Open();
				var oldChild = this.GetById(e.Id);
				if (oldChild == null)
				{
					_logger.Info($"Didn't find any child with id = {e.Id}");
						return null!;
				}
				using (var statement = new SQLiteCommand("update children set name = @name, cnp = @cnp where id = @id", connection))
				{
					statement.Parameters.Clear();
					statement.Parameters.AddWithValue("@name", e.Name);
					statement.Parameters.AddWithValue("@cnp", e.Cnp);
					statement.Parameters.AddWithValue("@id", e.Id);
					int rows = statement.ExecuteNonQuery();
					_logger.Info($"Affected rows: {rows}");
					return oldChild;
				}
			}
			catch (Exception ex)
			{
				_logger.Info($"Caught exception: " + ex);
				Console.WriteLine(ex.Message);
			}
			return null;
		}
	}
}

File: Semestru 4/MPP/lab/java/lab3/Persistence/src/main/java/org/example/repository/db/ChildDbRepository.java
package org.example.repository.db;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.domain.Child;
import org.example.repository.interf.IChildRepository;
import org.example.repository.utils.CommonUtils;

import java.sql.Connection;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;


public class ChildDbRepository implements IChildRepository {

    private CommonUtils Utils;
    private static final Logger logger = LogManager.getLogger();

    public ChildDbRepository(Properties properties) {
        logger.info("Initializing ChildDbRepository with properties: {}", properties);
        this.Utils = new CommonUtils(properties);
    }

    @Override
    public Iterable<Child> GetAllByAge(int age) {
        logger.traceEntry("Getting all of the children of age " + age);
        Connection connection = this.Utils.GetConnection();
        List<Child> children = new ArrayList<>();
        int year = LocalDateTime.now().getYear() % 100 - age;
        String yearString;
        if (year < 10)
            yearString = "_0" + year + "%";
        else yearString = "_" + year + "%";
        try (var statement = connection.prepareStatement("select * from children where cnp like ? ")) {
            statement.setString(1, yearString);
            try (var resultSet = statement.executeQuery()) {
                while (resultSet.next())
                    children.add(
                            (Child)new Child(
                                    resultSet.getString("name"),
                                    resultSet.getString("cnp")
                            ).SetId(resultSet.getInt("id"))
                    );
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return children;
    }

    @Override
    public Child FindById(Integer integer) {
        logger.traceEntry("Looking for child with id: {}", integer);
        Connection connection = this.Utils.GetConnection();
        Child child = null;
        try (var statement = connection.prepareStatement("select * from children where id = ?"))
        {
            statement.setInt(1, integer);
            try (var resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    child = (Child) new Child(resultSet.getString(2), resultSet.getString(3))
                            .SetId(resultSet.getInt(1));
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return child;
    }

    @Override
    public Iterable<Child> FindAll() {
        logger.traceEntry("Getting all of the children.");
        Connection connection = this.Utils.GetConnection();
        List<Child> children = new ArrayList<>();
        try (var statement = connection.prepareStatement("select * from children")) {
            try (var resultSet = statement.executeQuery()) {
                while (resultSet.next())
                    children.add(
                            (Child)new Child(
                                    resultSet.getString("name"),
                                    resultSet.getString("cnp")
                            ).SetId(resultSet.getInt("id"))
                    );

            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error" + e);
        }
        logger.traceExit();
        return children;
    }

    @Override
    public Child Save(Child entity) {
        logger.traceEntry("Saving child " + entity);
        Connection connection = this.Utils.GetConnection();
        try (var statement = connection.prepareStatement("insert into children (name, cnp) values (?, ?)"))
        {
            statement.setString(1, entity.GetName());
            statement.setString(2, entity.GetCNP());
            int result = statement.executeUpdate();
            try (var generatedKeys = statement.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    int generatedId = generatedKeys.getInt(1);
                    return (Child) entity.SetId(generatedId);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error " + e);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Child Update(Child entity) {
        logger.traceEntry("Updating child " + entity);
        Connection connection = this.Utils.GetConnection();
        Child oldChild = this.FindById(entity.GetId());
        try (var statement = connection.prepareStatement("update children set name = ?, cnp = ? where id = ?"))
        {
            statement.setString(1, entity.GetName());
            statement.setString(2, entity.GetCNP());
            statement.setInt(3, entity.GetId());
            statement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error " + e);
        }
        logger.traceExit();
        return oldChild;
    }

    @Override
    public Child Delete(Integer integer) {
        return null;
    }
}
